### PR TITLE
[Feature] Add MinkUNet segmentor

### PR DESCRIPTION
**Pull request by sunjiahao1999**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

1. Add minkunet voxelization
2. Add minkunet backbone
3. Add minkunet decode head
4. Add torchsparse optional backend

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
5. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
6. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
7. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by filaPro** [comment]
Btw, do we want to set different names to `resnet` and `unet` models based on `torchsparse` (in this PR), `minkowskiengine` (in `MinkResNet` class and I'm also thinking on adding `TD3D` segmentor with `MinkUNet`) and `spconv` (in `SparseUNet` class)?

**Comment by Xiangxu-0103** [comment]
@sunjiahao1999 Please pull the latest code and solve the conflicts.

**Comment by Xiangxu-0103** [comment]
> Btw, do we want to set different names to `resnet` and `unet` models based on `torchsparse` (in this PR), `minkowskiengine` (in `MinkResNet` class and I'm also thinking on adding `TD3D` segmentor with `MinkUNet`) and `spconv` (in `SparseUNet` class)?

Thanks for your reminder, we will take it into consideration.

**Comment by ZwwWayne** [comment]
> Btw, do we want to set different names to `resnet` and `unet` models based on `torchsparse` (in this PR), `minkowskiengine` (in `MinkResNet` class and I'm also thinking on adding `TD3D` segmentor with `MinkUNet`) and `spconv` (in `SparseUNet` class)?

I think we can do that.

**Comment by ZwwWayne** [comment]
We are also considering using same resnet/unet backbone that allow switching different sparse conv backends like mink/spconv/torchsparse

File: configs/minkunet/minkunet_cr0p5_8xb2-15e_semtickitti.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+_base_ = ['./minkunet_8xb2-15e_semtickitti.py']
+
+model = dict(
+    backbone=dict(
+        base_channels=16,
+        enc_channels=[16, 32, 64, 128],
```
Comment (Tai-Wang) [inline configs/minkunet/minkunet_cr0p5_8xb2-15e_semtickitti.py:6]: encode

File: configs/minkunet/minkunet_cr0p5_8xb2-15e_semtickitti.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+        dec_channels=[128, 64, 48, 48]),
```
Comment (Tai-Wang) [inline configs/minkunet/minkunet_cr0p5_8xb2-15e_semtickitti.py:7]: decode

File: mmdet3d/models/backbones/minkunet_backbone.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    from torchsparse.tensor import SparseTensor
+else:
+    SparseTensor = None
+
+
+@MODELS.register_module()
+class MinkUNetBackbone(BaseModule):
+    """MinkUNet backbone is the implementation of `4D Spatio-Temporal ConvNets.
+
+    <https://arxiv.org/abs/1904.08755>` with torchsparse backend.
```
Comment (Tai-Wang) [inline mmdet3d/models/backbones/minkunet_backbone.py:24]: `_

Comment (Tai-Wang) [reply mmdet3d/models/backbones/minkunet_backbone.py:24]: Please take the format of [this line](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/detectors/fcos_mono3d.py#L14) as the reference.

File: mmdet3d/models/backbones/minkunet_backbone.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            coors (Tensor): Coordinates in shape (N, 4),
+                the columns in the order of (batch_idx, z_idx, y_idx, x_idx).
+
+        Returns:
+            SparseTensor: backbone features.
+        """
+        x = torchsparse.SparseTensor(voxel_features, coors)
+        x = self.conv_input(x)
+        laterals = [x]
+        for enc in self.encoder:
```
Comment (Tai-Wang) [inline mmdet3d/models/backbones/minkunet_backbone.py:104]: enc -> encode_layer

File: mmdet3d/models/backbones/minkunet_backbone.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            x = enc(x)
+            laterals.append(x)
+        laterals = laterals[:-1][::-1]
+
+        dec_outs = []
```
Comment (Tai-Wang) [inline mmdet3d/models/backbones/minkunet_backbone.py:109]: dec_outs -> decode_outs

File: mmdet3d/models/backbones/minkunet_backbone.py (hunk 4 above)
Comment (Tai-Wang) [reply mmdet3d/models/backbones/minkunet_backbone.py:104]: also check other similar names

File: mmdet3d/models/data_preprocessors/data_preprocessor.py [hunk 6]
```diff
@@ -337,7 +338,7 @@ def voxelize(self, points: List[torch.Tensor],
         Args:
             points (List[Tensor]): Point cloud in one data batch.
             data_samples: (list[:obj:`Det3DDataSample`]): The annotation data
-                of every samples. Add voxel-wise annotation for segmentation.
+                of every samples. Add voxel-wise annotation forsegmentation.
```
Comment (Tai-Wang) [inline mmdet3d/models/data_preprocessors/data_preprocessor.py:341]: for segmentation

File: mmdet3d/models/decode_heads/minkunet_head.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+if IS_TORCHSPARSE_AVAILABLE:
+    from torchsparse import SparseTensor
+else:
+    SparseTensor = None
+
+
+@MODELS.register_module()
+class MinkUNetHead(Base3DDecodeHead):
+    """
```
Comment (Tai-Wang) [inline mmdet3d/models/decode_heads/minkunet_head.py:21]: Add a sentence description

File: mmdet3d/models/segmentors/minkunet.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    def __init__(self, **kwargs) -> None:
+        if not IS_TORCHSPARSE_AVAILABLE:
+            raise ImportError(
+                'Please follow `getting_started.md` to install Torchsparse.`'
+            )  # noqa: E501
+        super().__init__(**kwargs)
+
+    def loss(self, inputs: dict, data_samples: SampleList):
+        """"""
```
Comment (Tai-Wang) [inline mmdet3d/models/segmentors/minkunet.py:33]: Fix docstring

File: mmdet3d/models/segmentors/minkunet.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    Args:
+        data_preprocessor (dict, optional): Model preprocessing config
+            for processing the input data. it usually includes
+            ``to_rgb``, ``pad_size_divisor``, ``pad_val``,
+            ``mean`` and ``std``. Default to None.
+       init_cfg (dict, optional): the config to control the
+           initialization. Default to None.
+    """
+
+    def __init__(self, **kwargs) -> None:
```
Comment (Tai-Wang) [inline mmdet3d/models/segmentors/minkunet.py:25]: input arguments should be consistent with the docstring

File: configs/minkunet/minkunet_w16_8xb2-15e_semantickitti.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+_base_ = ['./minkunet_w32_8xb2-15e_semantickitti.py']
+
+model = dict(
+    backbone=dict(
```
Comment (JingweiZhang12) [inline configs/minkunet/minkunet_w16_8xb2-15e_semantickitti.py:4]: Please add README.md and describe the meaning of this config name and the wave of precision.

File: configs/minkunet/minkunet_w32_8xb2-15e_semantickitti.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    ),
+    dict(type='Pack3DDetInputs', keys=['points', 'pts_semantic_mask'])
+]
+
+train_dataloader = dict(
+    sampler=dict(seed=0), dataset=dict(dataset=dict(pipeline=train_pipeline)))
+
+lr = 0.24
+optim_wrapper = dict(
+    type='AmpOptimWrapper',
```
Comment (Xiangxu-0103) [inline configs/minkunet/minkunet_w32_8xb2-15e_semantickitti.py:37]: Why use `mixed precision training` here?

File: mmdet3d/models/data_preprocessors/data_preprocessor.py [hunk 12]
```diff
@@ -429,19 +457,49 @@ def get_voxel_seg(self, res_coors: torch.Tensor, data_sample: SampleList):
...
-                                                       res_coors, 'mean', True)
-            data_sample.gt_pts_seg.point2voxel_map = point2voxel_map
+        pts_semantic_mask = data_sample.gt_pts_seg.pts_semantic_mask
+        voxel_semantic_mask, _ = dynamic_scatter_3d(
+            F.one_hot(pts_semantic_mask.long()).float(),
+            res_coors.contiguous(), 'mean')
+        voxel_semantic_mask = torch.argmax(voxel_semantic_mask, dim=-1)
+        data_sample.gt_pts_seg.voxel_semantic_mask = voxel_semantic_mask
+
+    def ravel_hash(self, x: np.ndarray) -> np.ndarray:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/data_preprocessors/data_preprocessor.py:469]: Please add a docstring here

File: mmdet3d/models/data_preprocessors/data_preprocessor.py [hunk 13]
```diff
@@ -429,19 +457,49 @@ def get_voxel_seg(self, res_coors: torch.Tensor, data_sample: SampleList):
...
+            h *= xmax[k + 1]
+        h += x[:, -1]
+        return h
+
+    def sparse_quantize(self,
+                        coords,
+                        voxel_size: Union[float, Tuple[float, ...]] = 1,
+                        *,
+                        return_index: bool = False,
+                        return_inverse: bool = False) -> List[np.ndarray]:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/data_preprocessors/data_preprocessor.py:488]: Please add a docstring here.

File: mmdet3d/models/decode_heads/minkunet_head.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    Refer to `implementation code <https://github.com/mit-han-lab/spvnas>`_.
+
+    Args:
+        channels (int): The input channel of conv_seg.
+        num_classes (int): Number of classes.
+    """
+
+    def __init__(self, channels: int, num_classes: int, **kwargs) -> None:
+        super().__init__(channels, num_classes, **kwargs)
+        self.conv_seg = nn.Linear(channels, num_classes)
```
Comment (Xiangxu-0103) [inline mmdet3d/models/decode_heads/minkunet_head.py:32]: In the latest code, we may implement `build_conv_seg` instead of using `self.conv_seg=nn.Linear(channels, num_classes)`

File: mmdet3d/models/decode_heads/minkunet_head.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+    def loss(self, inputs: SparseTensor, data_samples: SampleList) -> dict:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/decode_heads/minkunet_head.py:34]: data_samples -> batch_data_samples

Comment (Xiangxu-0103) [inline mmdet3d/models/decode_heads/minkunet_head.py:34]: It seems that we just need to rewrite `_stack_batch_gt` and inherit `loss` from `Base3DDecodeHead`.

File: configs/minkunet/minkunet_w32_8xb2-15e_semantickitti.py (hunk 11 above)
Comment (sunjiahao1999) [reply configs/minkunet/minkunet_w32_8xb2-15e_semantickitti.py:37]: This is follow original implementation [here](https://github.com/mit-han-lab/spvnas/blob/69750e900d8687ac9fcc8e042b171cd1f6beffa1/configs/default.yaml#L3).

File: mmdet3d/models/decode_heads/minkunet_head.py [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+        Args:
+            inputs (SparseTensor): Features from backone.
+            batch_data_samples (List[:obj:`Det3DDataSample`]): The seg
+                data samples.
+
+        Returns:
+            List[Tensor]: The segmentation prediction mask of each batch.
+        """
+        seg_logits = self.forward(inputs)
+        seg_preds = seg_logits.argmax(dim=1)
```
Comment (Xiangxu-0103) [inline mmdet3d/models/decode_heads/minkunet_head.py:59]: We may generate `seg_preds` in segmentor. Here we just need to generate output logits for potential TTA model.

File: mmdet3d/models/segmentors/minkunet.py [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+
+    Args:
+        kwargs (dict): Arguments are the same as those in
+            :class:`EncoderDecoder3D`.
+    """
+
+    def __init__(self, **kwargs) -> None:
+        if not IS_TORCHSPARSE_AVAILABLE:
+            raise ImportError(
+                'Please follow `getting_started.md` to install Torchsparse.`')
```
Comment (Xiangxu-0103) [inline mmdet3d/models/segmentors/minkunet.py:27]: getting_started.md -> get_started.md

File: configs/minkunet/minkunet_w16_8xb2-15e_semantickitti.py [hunk 18]
```diff
@@ -0,0 +1,10 @@
...
+        base_channels=16,
+        encoder_channels=[16, 32, 64, 128],
+        decoder_channels=[128, 64, 48, 48]),
+    decode_head=dict(channels=48))
+
+randomness = dict(seed=1588147245)
```
Comment (JingweiZhang12) [inline configs/minkunet/minkunet_w16_8xb2-15e_semantickitti.py:10]: Please add a docstring above about the wave of performance.

**Review by JingweiZhang12 — APPROVED** [review]

File: configs/minkunet/minkunet_w32_8xb2-15e_semantickitti.py [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+        type='GlobalRotScaleTrans',
+        rot_range=[0., 6.28318531],
+        scale_ratio_range=[0.95, 1.05],
+        translation_std=[0, 0, 0],
+    ),
+    dict(type='Pack3DDetInputs', keys=['points', 'pts_semantic_mask'])
+]
+
+train_dataloader = dict(
+    sampler=dict(seed=0), dataset=dict(dataset=dict(pipeline=train_pipeline)))
```
Comment (Tai-Wang) [inline configs/minkunet/minkunet_w32_8xb2-15e_semantickitti.py:27]: Is the seed necessary here?

File: mmdet3d/models/data_preprocessors/data_preprocessor.py [hunk 20]
```diff
@@ -429,19 +456,57 @@ def get_voxel_seg(self, res_coors: torch.Tensor, data_sample: SampleList):
...
+        return h
+
+    def sparse_quantize(self,
+                        coords: np.ndarray,
+                        return_index: bool = False,
+                        return_inverse: bool = False) -> List[np.ndarray]:
+        """Sparse Quantization for voxel coordinates used in Minkunet.
+
+        Args:
+            coords (np.ndarray): The voxel coordinates of points, Nx3.
```
Comment (Tai-Wang) [inline mmdet3d/models/data_preprocessors/data_preprocessor.py:497]: need to supplement the docstring for other two arguments

File: mmdet3d/models/decode_heads/minkunet_head.py [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+        return seg_logit_list
+
+    def forward(self, x: SparseTensor) -> Tensor:
+        """Forward function.
+
+        Args:
+            x (SparseTensor): Features from backbone.
+
+        Returns:
+            output (Tensor): Segmentation map of shape [N, C].
```
Comment (Tai-Wang) [inline mmdet3d/models/decode_heads/minkunet_head.py:76]: fix the format

File: mmdet3d/models/data_preprocessors/data_preprocessor.py [hunk 22]
```diff
@@ -429,19 +456,57 @@ def get_voxel_seg(self, res_coors: torch.Tensor, data_sample: SampleList):
...
             data_sample: (:obj:`Det3DDataSample`): The annotation data of
-                every samples. Add voxel-wise annotation forsegmentation.
+                every samples. Add voxel-wise annotation for segmentation.
         """
+        pts_semantic_mask = data_sample.gt_pts_seg.pts_semantic_mask
+        voxel_semantic_mask, _ = dynamic_scatter_3d(
+            F.one_hot(pts_semantic_mask.long()).float(),
+            res_coors.contiguous(), 'mean')
+        voxel_semantic_mask = torch.argmax(voxel_semantic_mask, dim=-1)
+        data_sample.gt_pts_seg.voxel_semantic_mask = voxel_semantic_mask
```
Comment (Xiangxu-0103) [inline mmdet3d/models/data_preprocessors/data_preprocessor.py:466]: Whether this modification will cause any bc-breaking for cylinder3D?

File: mmdet3d/models/segmentors/minkunet.py [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+                `gt_pts_seg`. Defaults to None.
+
+        Returns:
+            Tensor: Forward output of model without any post-processes.
+        """
+        x = self.extract_feat(batch_inputs_dict)
+        return self.decode_head.forward(x)
+
+    def extract_feat(self, batch_inputs_dict: dict) -> Tuple[Tensor]:
+        """Extract features from voxels."""
```
Comment (ZCMax) [inline mmdet3d/models/segmentors/minkunet.py:98]: add args and returns here

File: tests/test_models/test_decode_heads/test_minkunet_head.py [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+
+from mmdet3d.models.decode_heads import MinkUNetHead
+from mmdet3d.structures import Det3DDataSample, PointData
+
+
+class TestMinkUNetHead(TestCase):
+
+    def test_minkunet_head_loss(self):
+        """Tests PAConv head loss."""
+
```
Comment (ZCMax) [inline tests/test_models/test_decode_heads/test_minkunet_head.py:16]: should we add ?
```
if not torch.cuda.is_available():
         pytest.skip('test requires GPU and torch+cuda')
```

File: tests/test_models/test_segmentor/test_minkunet.py [hunk 25]
```diff
@@ -0,0 +1,45 @@
+import unittest
```
Comment (ZCMax) [inline tests/test_models/test_segmentor/test_minkunet.py:1]: please add # Copyright (c) OpenMMLab. All rights reserved.

**Comment by ZCMax** [comment]
minor comments to be fixed and then I'll approve the PR

**Review by Xiangxu-0103 — APPROVED** [review]

**Review by ZCMax — APPROVED** [review]
